src/neo/SmartContract/ApplicationDebugger.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (C) 2015-2021 The Neo Project.
+// 
+// The neo is free software distributed under the MIT software license, 
+// see the accompanying file LICENSE in the main directory of the
+// project or http://www.opensource.org/licenses/mit-license.php 
+// for more details.
+// 
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using Neo.Network.P2P.Payloads;
+using Neo.Persistence;
+using Neo.VM;
+
+namespace Neo.SmartContract
+{
+    public class ApplicationDebugger : ApplicationEngine
+    {
+        protected ApplicationDebugger(TriggerType trigger, IVerifiable container, DataCache snapshot, Block persistingBlock, ProtocolSettings settings, long gas, Diagnostic diagnostic) : base(trigger, container, snapshot, persistingBlock, settings, gas, diagnostic)
+        {
+        }
+
+        public static new ApplicationDebugger Create(TriggerType trigger, IVerifiable container, DataCache snapshot, Block persistingBlock = null, ProtocolSettings settings = null, long gas = TestModeGas, Diagnostic diagnostic = null)
+        {
+            return new ApplicationDebugger(trigger, container, snapshot, persistingBlock, settings, gas, diagnostic);
+        }
+
+        public new VMState State
+        {
+            get
+            {
+                return base.State;
+            }
+            set
+            {
+                base.State = value;
```

Comment:
Why do you need to set `State`?

Reply:
To allow VMState.BREAK

Reply:
I think this can be done by https://github.com/neo-project/neo-vm/blob/master/src/neo-vm/Debugger.cs

Reply:
The Debugger does not have abundant application-level features like CurrentScriptHash, gasConsumed, etc.

Reply:
You can wrap it, right?

Reply:
In order to build an ApplicationEngine-level debugger from neo-vm/Debugger, there have to be lots of duplicate codes. I can abort this pull request and put ApplicationDebugger only where I use it.